sklearn/semi_supervised/tests/test_self_training.py

```diff
@@ -0,0 +1,189 @@
+# License: BSD 3 clause
+
+# load the iris dataset and randomly permute it
+iris = load_iris()
+X_train, X_test, y_train, y_test = train_test_split(iris.data,
+                                                    iris.target,
+                                                    random_state=0)
+
+n_labeled_samples = 50
+
+y_train_missing_labels = y_train.copy()
+y_train_missing_labels[n_labeled_samples:] = -1
+mapping = {0: 'A', 1: 'B', 2: 'C', -1: '-1'}
+y_train_missing_strings = np.vectorize(mapping.get)(
+    y_train_missing_labels).astype(object)
+y_train_missing_strings[y_train_missing_labels == -1] = -1
+
+lb = LabelBinarizer()
+y_train_missing_dummies = lb.fit_transform(y_train_missing_labels)
+
+
+def test_missing_predict_proba():
+    # Check that an error is thrown if predict_proba is not implemented
+    base_classifier = SVC(gamma="scale")
+    self_training = SelfTrainingClassifier(base_classifier)
+    message = "base_classifier (SVC) should implement predict_proba!"
+    assert_raise_message(ValueError, message, self_training.fit, X_train,
+                         y_train)
+
+
+def test_none_classifier():
+    st = SelfTrainingClassifier(None)
+    msg = "base_classifier cannot be None"
+    assert_raise_message(ValueError, msg, st.fit, X_train,
+                         y_train_missing_labels)
+
+
+@pytest.mark.parametrize("max_iter, threshold",
+                         [(-1, 1.0), (-100, -2), (-10, 10)])
+def test_invalid_params(max_iter, threshold):
+    # Test negative iterations
+    base_classifier = SVC(gamma="scale", probability=True)
+    st = SelfTrainingClassifier(base_classifier, max_iter=max_iter)
+    message = "max_iter must be >= 0 or None, got"
+    assert_raise_message(ValueError, message, st.fit, X_train, y_train)
+
+    base_classifier = SVC(gamma="scale", probability=True)
+    st = SelfTrainingClassifier(base_classifier, threshold=threshold)
+    message = "threshold must be in [0,1)"
+    assert_raise_message(ValueError, message, st.fit, X_train, y_train)
+
+
+@pytest.mark.parametrize("base_classifier",
+                         [DummyClassifier(random_state=0),
+                          DecisionTreeClassifier(random_state=0),
+                          KNeighborsClassifier(),
+                          SVC(gamma="scale", probability=True,
+                              random_state=0)])
+@pytest.mark.parametrize("max_iter", [1, 50, 100])
+@pytest.mark.parametrize("threshold", [0.0, 0.5, 0.9])
+def test_classification(base_classifier, max_iter, threshold):
+    # Check classification for various parameter settings.
+    # Also assert that predictions for strings and numerical labels are equal.
+    # Also test for multioutput classification
+    st = SelfTrainingClassifier(base_classifier, max_iter=max_iter,
+                                threshold=threshold)
+    st.fit(X_train, y_train_missing_labels)
+    pred = st.predict(X_test)
+    proba = st.predict_proba(X_test)
+
```

Comment:
we should also test output attributes, such as n_iter_ being consistent with max_iter, and y_labeled_iter_ being consistent with n_iter_ and y_label_ being consistent with y_labeled_iter_

Reply:
> y_label_ being consistent with y_labeled_iter_
Do you mean the shapes?

Reply:
> > y_label_ being consistent with y_labeled_iter_

> Do you mean the shapes?

And the -1 mask